datalake/src/main/java/com/sequenceiq/datalake/service/SdxEventsService.java

```diff
@@ -0,0 +1,162 @@
+import java.util.Comparator;
+import java.util.List;
+import java.util.stream.Collectors;
+
+import javax.inject.Inject;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.data.domain.Page;
+import org.springframework.data.domain.PageRequest;
+import org.springframework.data.domain.Sort;
+import org.springframework.stereotype.Service;
+
+import com.sequenceiq.cloudbreak.api.endpoint.v4.events.EventV4Endpoint;
+import com.sequenceiq.cloudbreak.api.endpoint.v4.events.responses.CloudbreakEventV4Response;
+import com.sequenceiq.cloudbreak.auth.crn.Crn;
+import com.sequenceiq.cloudbreak.auth.crn.CrnParseException;
+import com.sequenceiq.cloudbreak.common.exception.BadRequestException;
+import com.sequenceiq.cloudbreak.structuredevent.event.CloudbreakEventService;
+import com.sequenceiq.cloudbreak.structuredevent.event.StructuredEventType;
+import com.sequenceiq.cloudbreak.structuredevent.event.cdp.CDPOperationDetails;
+import com.sequenceiq.cloudbreak.structuredevent.event.cdp.CDPStructuredEvent;
+import com.sequenceiq.cloudbreak.structuredevent.event.cdp.CDPStructuredNotificationEvent;
+import com.sequenceiq.cloudbreak.structuredevent.service.db.CDPStructuredEventDBService;
+import com.sequenceiq.datalake.entity.SdxCluster;
+import com.sequenceiq.datalake.repository.SdxClusterRepository;
+import com.sequenceiq.datalake.service.sdx.SdxService;
+
+@Service
+public class SdxEventsService {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(SdxEventsService.class);
+
+    @Inject
+    private SdxClusterRepository sdxClusterRepository;
+
+    @Inject
+    private SdxService sdxService;
+
+    @Inject
+    private CDPStructuredEventDBService cdpStructuredEventDBService;
+
+    @Inject
+    private EventV4Endpoint eventV4Endpoint;
+
+    public List<CDPStructuredEvent> getDatalakeAuditEvents(String environmentCrn, List<StructuredEventType> types, Integer page, Integer size) {
+        List<CDPStructuredEvent> dlEvents;
+        List<CDPStructuredEvent> cbEvents;
+
+        PageRequest pageable = PageRequest.of(page, size, Sort.by("timestamp").descending());
+        SdxCluster sdxCluster = getDatalake(environmentCrn);
+        if (sdxCluster == null) {
+            LOGGER.error("Datalake not found for environment with Crn:{}", environmentCrn);
+            return List.of();
+        }
+
+        List<String> datalakeCrns = getDatalakeCrns(environmentCrn);
+        dlEvents = retrieveDatalakeServiceEvents(types, datalakeCrns, pageable);
+        cbEvents = retrieveCloudbreakServiceEvents(sdxCluster, page, size);
+
+        List<CDPStructuredEvent> combinedEvents = new ArrayList<>();
+        combinedEvents.addAll(cbEvents);
+        combinedEvents.addAll(dlEvents);
+
+        if (combinedEvents.isEmpty()) {
+            LOGGER.info("No events from datalake and cloudbreak service");
+            return List.of();
+        }
+
+        // Sort and Filter the events out based on page size
```

Comment:
can we put it into a method like sortAndFilterBasedOnPageSize()? comments can be easily out of date because people forget to update it

Reply:
ok

Reply:
done